pkg/utils/host/host.go

```diff
@@ -80,16 +97,130 @@ func init() {
+	if HostRoot != "/host" || os.Getenv("KUBERNETES_PORT") == "" {
+		return nil
+	}
+
+	c, err := capability.NewPid2(0)
+	if err != nil {
 		return err
 	}
+	err = c.Load()
+	if err != nil {
+		return err
+	}
+	if c.Get(capability.EFFECTIVE, capability.CAP_SYS_ADMIN) {
+		// if we already have CAP_SYS_ADMIN, we don't need a workaround
+		return nil
+	}
+
+	_, err = os.Stat("/host/run/systemd/private")
+	if err != nil {
+		// if the host does not use systemd, we cannot use this workaround
+		return nil
+	}
+
+	// systemdDbus.NewSystemdConnectionContext() hardcodes the path to the
+	// systemd socket to /run/systemd/private. We need to make sure that this
+	// path exists (if the /run:/run mount was set up correctly). If it doesn't
+	// exist, we create the symlink to /host/run/systemd/private.
+	_, err = os.Stat("/run/systemd/private")
+	if errors.Is(err, os.ErrNotExist) {
+		err := os.MkdirAll("/run/systemd", 0o755)
+		if err != nil {
+			return err
+		}
+
+		err = os.Symlink("/host/run/systemd/private", "/run/systemd/private")
+		if err != nil {
+			return fmt.Errorf("linking /run/systemd/private: %w", err)
+		}
+	} else if err != nil {
+		return fmt.Errorf("statting /run/systemd/private: %w", err)
+	}
+
+	conn, err := systemdDbus.NewSystemdConnectionContext(context.TODO())
+	if err != nil {
+		return fmt.Errorf("connecting to systemd: %w", err)
+	}
+	defer conn.Close()
+
+	runID := uuid.New().String()[:8]
+	unitName := fmt.Sprintf("kubectl-debug-ig-%s.service", runID)
+
+	exit := make(chan os.Signal, 1)
+	signal.Notify(exit, syscall.SIGINT, syscall.SIGTERM)
+
+	statusChan := make(chan string, 1)
+	cmd := []string{
+		fmt.Sprintf("/proc/%d/root/usr/bin/ig", os.Getpid()),
+	}
+	cmd = append(cmd, os.Args[1:]...)
+	properties := []systemdDbus.Property{
+		systemdDbus.PropDescription("Inspektor Gadget via kubectl debug"),
+		// Type=oneshot ensures that StartTransientUnitContext will only return "done" when the job is done
+		systemdDbus.PropType("oneshot"),
+		{
+			Name:  "StandardInputFileDescriptor",
+			Value: dbus.MakeVariant(dbus.UnixFD(unix.Stdin)),
+		},
+		{
+			Name:  "StandardOutputFileDescriptor",
+			Value: dbus.MakeVariant(dbus.UnixFD(unix.Stdout)),
+		},
+		{
+			Name:  "StandardErrorFileDescriptor",
+			Value: dbus.MakeVariant(dbus.UnixFD(unix.Stderr)),
+		},
+		systemdDbus.PropExecStart(cmd, true),
+	}
+
+	isTerminal := term.IsTerminal(int(os.Stdin.Fd())) || term.IsTerminal(int(os.Stdout.Fd())) || term.IsTerminal(int(os.Stderr.Fd()))
+	if isTerminal && os.Getenv("TERM") != "" {
+		p := systemdDbus.Property{
+			Name:  "Environment",
+			Value: dbus.MakeVariant([]string{"TERM=" + os.Getenv("TERM")}),
+		}
+		properties = append(properties, p)
+	}
+
+	_, err = conn.StartTransientUnitContext(context.TODO(),
+		unitName, "fail", properties, statusChan)
+	if err != nil {
+		return fmt.Errorf("starting transient unit %q: %w", unitName, err)
+	}
+
+	select {
+	case s := <-statusChan:
+		log.Infof("systemd unit %q returned %q", unitName, s)
```

Comment:
I don't see any of these logs. It'd be nice to show to the user that this workaround is being used.

Reply:
This one is printed if the systemd unit returns by itself first. For example, with `--timeout 5`. If you press ctrl-c, it will go through the second `case`.

I'll add a debug to tell when the workaround is used.

But I think it should be a `log.Debugf` (only printed with `--verbose`): if the command returns json data, the log would still be parsable.

Reply:
Ah true...

When using the `-t` option of `kubectl debug`, it creates a tty, so stdout and stderr are both the same file descriptor to the tty. I guess in that case users should not use `-t` if they want to parse the json... and we should check if that works fine.